### PR TITLE
feat: pagination + metadata for csv and excel readers [TCTC-1810]

### PR DESCRIPTION
**Pull request by luc-leonard**
## Change Summary

Split https://github.com/ToucanToco/peakina/pull/69 in two PR.

* with this PR the slicing is no longer done by Laputa, but directly by peakina.

* this PR enables such slicing for csv and excel files. 

* It does nothing to improve the loading performance of excel files.

* include a small change from https://github.com/ToucanToco/peakina/pull/69 : the HttpFetcher now preserve the remote file extension.

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable

### CODE REVIEW

File: peakina/datasource.py [hunk 1]
```diff
@@ -107,12 +107,6 @@ def _get_single_df(
         finally:
             stream.close()
 
-        # In case of sheets, the df can be a dictionary
-        if kwargs.get("sheet_name", NOTSET) is None:
-            for sheet_name, _df in df.items():
-                _df["__sheet__"] = sheet_name
-            df = pd.concat(df.values(), sort=False)
-
```
Comment (luc-leonard) [inline peakina/datasource.py:115]: this is now part of the excel reader

File: peakina/helpers.py [hunk 2]
```diff
@@ -39,24 +42,32 @@ class TypeInfos(NamedTuple):
 
 
 SUPPORTED_FILE_TYPES = {
-    "csv": TypeInfos(["text/csv", "text/tab-separated-values"], pd.read_csv),
+    "csv": TypeInfos(
+        ["text/csv", "text/tab-separated-values"],
+        read_csv,
+        ["preview_offset", "preview_nrows", "skiprows"],
```
Comment (luc-leonard) [inline peakina/helpers.py]: preview_offset and preview_nrows are kwargs that are available on all readers

File: peakina/readers/csv.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            preview_offset = 1  # skip header
+        chunks = pd.read_csv(
+            filepath,
+            sep=sep,
+            header=None,
+            names=_extract_columns(filepath, encoding, sep),
+            keep_default_na=keep_default_na,
+            encoding=encoding,
+            nrows=preview_nrows,
+            skiprows=lambda idx: idx < preview_offset + (skiprows or 0),
```
Comment (notaustil) [inline peakina/readers/csv.py]: > **skiprows:list-like, int or callable, optional**
>   Line numbers to skip (0-indexed) or number of lines to skip (int) at the start of the file.
>    If callable, the callable function will be evaluated against the row indices, returning True if the row should be skipped and False otherwise. An example of a valid callable argument would be lambda x: x in [0, 2].

Why using a lambda and not "just" the number of row to skip from the start ? Does this have something to do with other options ? Does this have any impact on performance when preview pages at the end ?

File: peakina/readers/excel.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def excel_meta(filepath: str, datasource: "DataSource") -> dict:  # noqa: F821
+    """
+    Returns a dictionary with the meta information of the excel file.
+    """
+    kwargs = datasource.reader_kwargs
+    kwargs.pop("preview_offset", None)
+    kwargs.pop("preview_nrows", None)
+    df = pd.read_excel(filepath, **datasource.reader_kwargs)
```
Comment (notaustil) [inline peakina/readers/excel.py]: So in some cases, we'll be reading the Excel files twice ? making twice as slow ?

Comment (notaustil) [reply peakina/readers/excel.py]: Precision : we have two `pd.read_excel`, one for meta, one for content

Comment (notaustil) [reply peakina/readers/excel.py]: KK, so this is a perf regression for Excel file, but it's okay as long as we're sure to put the work necessary to merge https://github.com/ToucanToco/peakina/pull/69 before next stable release.

File: peakina/datasource.py [hunk 5]
```diff
@@ -91,7 +91,7 @@ def _get_single_df(
         allowed_params = get_reader_allowed_params(filetype)
 
         # Check encoding
-        encoding = kwargs.get("encoding")
+        encoding = kwargs.get("encoding", "utf-8")
```
Comment (raphaelvignes) [inline peakina/datasource.py:93]: :+1:

File: peakina/datasource.py (hunk 1 above)
Comment (raphaelvignes) [reply peakina/datasource.py:115]: :+1: Thanks for the explanation.

File: peakina/helpers.py (hunk 2 above)
Comment (raphaelvignes) [reply peakina/helpers.py]: so we keep them for `read_json` & `read_xml` to have the same signature ?

File: peakina/io/http/http_fetcher.py [hunk 6]
```diff
@@ -22,7 +23,10 @@ def __init__(self, *, verify: bool = True, **kwargs: Any) -> None:
 
     def open(self, filepath: str) -> IO[bytes]:
         r = self.pool_manager.request("GET", filepath, preload_content=False, **self.extra_kwargs)
-        ret = tempfile.NamedTemporaryFile(suffix=".httptmp")
+        extension = os.path.splitext(filepath)[1]
+        if "?" in extension:  # remove query string
```
Comment (raphaelvignes) [inline peakina/io/http/http_fetcher.py]: Pourquoi est-ce qu'on doit appliquer cette logique ?

File: peakina/readers/excel.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        for sheet_name, _df in df.items():
+            _df["__sheet__"] = sheet_name
+        df = pd.concat(df.values(), sort=False)
+
+    if preview_offset is not None and preview_nrows is not None:
+        return df[preview_offset: preview_offset + preview_nrows]
+    return df
+
+
+def excel_meta(filepath: str, datasource: "DataSource") -> dict:  # noqa: F821
```
Comment (raphaelvignes) [inline peakina/readers/excel.py]: Is it mandatory to call `pd.read_excel` to get only nrows & sheetnames ? 
With `pyopenxl` you can retrieve it like this: 

https://stackoverflow.com/questions/38285364/python-openpyxl-read-only-mode-returns-a-different-row-count

**Review by raphaelvignes — COMMENTED** [review]
OK for this PR but regarding this comment: https://github.com/ToucanToco/peakina/pull/77#discussion_r809796123 
It seemed to me that it was possible to obtain the metadata without loading the whole file

File: peakina/io/http/http_fetcher.py (hunk 6 above)
Comment (luc-leonard) [reply peakina/io/http/http_fetcher.py]: Parce que openpyxl et xlrd se basent sur l'extension pour determiner le type du fichier :yum:

File: peakina/readers/excel.py (hunk 7 above)
Comment (luc-leonard) [reply peakina/readers/excel.py]: Oui, mais faut reimplementer la logique de fallback sur xldr pour les .xls

Comment (raphaelvignes) [reply peakina/readers/excel.py]: OK

Comment (notaustil) [reply peakina/readers/excel.py]: We're not using `pyopenxl` in this PR, so it's not perfect but fine as long as we fix it in another one.

File: Makefile [hunk 8]
```diff
@@ -19,7 +19,6 @@ format:
 
 .PHONY: lint
 lint:
-	poetry run flake8 peakina tests
```
Comment (PrettyWood) [inline Makefile]: Why did you remove that?

File: peakina/io/http/http_fetcher.py (hunk 6 above)
Comment (PrettyWood) [reply peakina/io/http/http_fetcher.py]: Why did you modify the http fetcher? How is it related to csv and excel? 🤔

File: peakina/readers/csv.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+import pandas as pd
+
+# The chunksize value for previews
+PREVIEW_CHUNK_SIZE = 1024
+
+
+def _extract_columns(filepath: str, encoding: str, sep: str) -> List[str]:
+    with open(filepath, buffering=10000, encoding=encoding) as ff:
+        return ff.readline().replace("\n", "").split(sep)
```
Comment (PrettyWood) [inline peakina/readers/csv.py]:
```suggestion
        return ff.readline().rstrip("\n").split(sep)
```
Probably safer and more explicit

File: peakina/readers/csv.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    preview_offset: Optional[int] = None,
+    preview_nrows: Optional[int] = None,
+    chunksize: Optional[int] = None,
+    nrows: int = 500,
+    error_bad_lines: bool = False,
+    skiprows: Optional[int] = None,
+) -> pd.DataFrame:
+    """
+    The read_csv method is able to make a preview by reading on chunks
+
```
Comment (PrettyWood) [inline peakina/readers/csv.py]:
```suggestion
```

File: peakina/helpers.py (hunk 2 above)
Comment (PrettyWood) [reply peakina/helpers.py]: If you add the kwargs in the signature, then you don't need to add them here normally

File: peakina/readers/csv.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    """
+    if preview_nrows is not None and preview_offset is not None:
+        if preview_offset == 0:
+            preview_offset = 1  # skip header
```
Comment (PrettyWood) [inline peakina/readers/csv.py]: Don't we want to always skip header (meaning always add +1) ?
If I put `offset=10` I expect the first 10 rows to be skipped not only the first 9

File: peakina/readers/csv.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            header=None,
+            names=_extract_columns(filepath, encoding, sep),
+            keep_default_na=keep_default_na,
+            encoding=encoding,
+            nrows=preview_nrows,
+            skiprows=preview_offset + (skiprows or 0),
+            chunksize=PREVIEW_CHUNK_SIZE,
+            error_bad_lines=error_bad_lines,
+        )
+        return next(chunks)
```
Comment (PrettyWood) [inline peakina/readers/csv.py:44]: 👍

File: peakina/readers/csv.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        nrows=nrows,
+        sep=sep,
+        chunksize=chunksize,
+        encoding=encoding,
+        keep_default_na=keep_default_na,
+        skiprows=skiprows,
+    )
+
+
+def _line_count(filename: str) -> int:
```
Comment (PrettyWood) [inline peakina/readers/csv.py]: 👍

**Review by PrettyWood — CHANGES_REQUESTED** [review]

File: Makefile (hunk 8 above)
Comment (luc-leonard) [reply Makefile]: Because flake8 and black disagree about formatting and seem to be mutually exclusive

flake8 wants `df[begin: end]`
black wants `df[begin:end]`

File: peakina/readers/csv.py (hunk 11 above)
Comment (luc-leonard) [reply peakina/readers/csv.py]: :+1:

File: peakina/readers/csv.py (hunk 9 above)
Comment (luc-leonard) [reply peakina/readers/csv.py]: > Why did you modify the http fetcher? How is it related to csv and excel? thinking

because openpyxl and xlrd are using the file extension to know to know if this is an xls or an xlsx file. It is not required by this PR thouhgh, so it can be removed for the time being.

File: Makefile (hunk 8 above)
Comment (PrettyWood) [reply Makefile]: Yes you're right it should be ignored in `flake8` config. I added this in `main`. Can you please merge or rebase?

File: peakina/readers/excel.py (hunk 7 above)
Comment (Sanix-Darker) [reply peakina/readers/excel.py]: 👍🏾

File: peakina/helpers.py [hunk 14]
```diff
@@ -37,18 +39,23 @@ class TypeInfos(NamedTuple):
 # For files without MIME types, we make fake MIME types based on detected extension
 CUSTOM_MIMETYPES = {".parquet": "peakina/parquet"}
 
+EXTRA_PEAKINA_READER_KWARGS = ["preview_offset", "preview_nrows"]
 
 SUPPORTED_FILE_TYPES = {
```
Comment (PrettyWood) [inline peakina/helpers.py:49]: I think it will be better to change this in the future into `CsvReader`, `ExcelReader`, ... with class methods

**Review by PrettyWood — APPROVED** [review]

File: peakina/readers/csv.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+            buf = read_f(buf_size)
+
+        return lines
+
+
+def csv_meta(
+    filepath_or_buffer: "FilePathOrBuffer", reader_kwrgs: Dict[str, Any]
+) -> Dict[str, Any]:
+    return {
+        "nrows": _line_count(filepath_or_buffer),
```
Comment (PrettyWood) [inline peakina/readers/csv.py]: Let's do `total_rows` and `df_rows` for example

File: peakina/readers/excel.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    from peakina.helpers import EXTRA_PEAKINA_READER_KWARGS
+
+    excel_reader_kwargs = {
+        k: v for k, v in reader_kwargs.items() if k not in EXTRA_PEAKINA_READER_KWARGS
+    }
+    excel_file = pd.ExcelFile(filepath_or_buffer)
+    df = pd.read_excel(excel_file, **excel_reader_kwargs)
+    return {
+        "sheetnames": excel_file.sheet_names,
+        "nrows": df.shape[0],
```
Comment (PrettyWood) [inline peakina/readers/excel.py]: what if `df` is a dictionary? (multiple sheets)

**Review by PrettyWood — CHANGES_REQUESTED** [review]

File: peakina/readers/csv.py [hunk 17]
```diff
@@ -51,23 +51,24 @@ def read_csv(
...
 
-        return lines
+    df_rows = pd.read_csv(
+        filepath_or_buffer,
+        **reader_kwrgs,
+        skiprows=range(1, preview_offset + 1) if preview_offset else None,
+        nrows=preview_nrows,
+    ).shape[0]
 
+    total_rows = pd.read_csv(filepath_or_buffer, **reader_kwrgs).shape[0]
```
Comment (PrettyWood) [inline peakina/readers/csv.py]: `_line_count` has been created to count very fast the total of rows of a huge csv!
This was the whole point because we don't want to use `read_excel` (which will load everything in memory) just to get metadata!

Comment (PrettyWood) [reply peakina/readers/csv.py]: I would have done something like
```py
total_rows = _line_count(filepath_or_buffer)

preview_offset = reader_kwargs.pop("preview_offset", 0)
preview_nrows = reader_kwargs.pop("preview_nrows", None)
if preview_nrows is not None:
    df_rows = min(total_rows, preview_nrows - preview_offset)
else:
    df_rows = total_rows
```
with a test
- where we ask all the rows of a file of 4 rows and `df_rows` must equal 4
- where we ask 5 rows of a file of 4 rows and `df_rows` must equal 4
- where we ask 3 rows of a file of 4 rows and `df_rows` must equal 3

Wouldn't that work?

**Review by PrettyWood — CHANGES_REQUESTED** [review]

File: peakina/readers/csv.py (hunk 17 above)
Comment (Sanix-Darker) [reply peakina/readers/csv.py]: thanks for updates, i replicated these tests scenarios here : https://github.com/ToucanToco/peakina/pull/77/commits/5d34fe5c7d30125798e8638c3d1e630519a9604c

File: peakina/readers/excel.py (hunk 16 above)
Comment (Sanix-Darker) [reply peakina/readers/excel.py]: Please check how i handled that here : https://github.com/ToucanToco/peakina/pull/77/commits/9061b366d26a758becf2e1b39f4772cee5513eca  and let me know what you think about it !

File: peakina/readers/csv.py [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+    preview_nrows = reader_kwrgs.pop("preview_nrows", None)
+    if preview_nrows is not None:
+        if preview_nrows <= preview_offset:
+            df_rows = preview_nrows
+        else:
+            df_rows = min(total_rows, preview_nrows - preview_offset)
+    else:
+        df_rows = total_rows - preview_offset
+
+    return {"total_rows": _line_count(filepath_or_buffer), "df_rows": df_rows}
```
Comment (luc-leonard) [inline peakina/readers/csv.py]: image

These options will affect the number of lines in the dataset. They should be taken into account when counting rows.

**Review by PrettyWood — APPROVED** [review]
